elastic/elastic-agent
# Send package.version to components

## What does this PR do?

The agent now includes AgentInfo alongside the connection information during component startup. It also stores the BuildHash sent by the components as part of the ComponentVersionInfo during check-in.

To facilitate testing, an info.Agent interface has been created to abstract the implementation of info.AgentInfo. This abstraction allows for the creation of mocks, enabling testing scenarios where using `info.NewAgentInfo` is impractical. This is because `info.NewAgentInfo` relies on the agent vault, which, on Mac systems, is the system's keychain. Accessing the keychain requires root permissions, which are not available during testing.

## Why is it important?

The components should report the package version instead of their own.

## Checklist





## How to test this PR locally

With a beats build containing https://github.com/elastic/beats/pull/37553:
 - build the agent
 - change the package.version file to contain a version with extra build information, e.g.: `8.13.0+build20131123`
 - enroll it to fleet
 - the events sent by the beats should have the `agent.version`, `elastic_agent.version` and `metadata.version` set as the version defined on the `package.version` file.
 - in the `state.yaml` from the agent diagnostics, for each component, there should be:
   - no `version` field in the `version_info`
   - a new `version_info.build_hash` with the beats build hash

## Related issues

- N/A

## Related PRs
 - https://github.com/elastic/elastic-agent-client/pull/93
 - https://github.com/elastic/beats/pull/37553


## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...
